### ISSUE TITLE
Volspotconnect2 - Multi-User Device not working

### ISSUE THREAD

**Issue by Squixx** [balbuze/volumio-plugins#135]
Since the last spotify update ( not sure, started out of nowhere yesterday)

Multi-User Device makes volumio not selectable from spotify (android AND windows desktop)

Could be related to: https://github.com/librespot-org/librespot/issues/181 ?
 
but im not sure.

Logging in directly to my spotify causes the device to reappear directly.

**Comment by balbuze** [comment]
Please install the version my repo https://github.com/balbuze/volumio-plugins/raw/master/plugins/music_service/volspotconnect2/volspotconnect2.zip
It should solve it

**Comment by Squixx** [comment]
It seems to only start working after I configure an account, when saving loging details. connecting and then switching back to multi-user it seems to remain active..

Underlying log seems to indicate no session is being broadcatsted before user auth is done.



`sudo journalctl -f | grep 'librespot\|SpCon'
Mar 03 16:42:58 volumio volumio[810]: info: [SpConDebug] Killing Volumio State
Mar 03 16:42:58 volumio volumio[810]: info: [SpConDebug] SpotConn is done!
Mar 03 16:42:59 volumio volumio[2596]: INFO:librespot: librespot c1128f9 (2018-02-28). Built on 2018-02-28. Build ID: B698bwAU
Mar 03 16:43:21 volumio volumio[810]: info: [SpConDebug] Killing Volumio State
Mar 03 16:43:21 volumio volumio[810]: info: [SpConDebug] SpotConn is done!
Mar 03 16:43:22 volumio volumio[2756]: INFO:librespot: librespot c1128f9 (2018-02-28). Built on 2018-02-28. Build ID: B698bwAU
Mar 03 16:43:22 volumio volumio[2756]: INFO:librespot_core::session: Connecting to AP "gew1-accesspoint-b-dq9r.ap.spotify.com:4070"
Mar 03 16:43:24 volumio volumio[2756]: INFO:librespot_core::session: Authenticated as "11669317" !
Mar 03 16:43:24 volumio volumio[2756]: INFO:librespot::audio_backend::alsa: Using alsa sink
Mar 03 16:43:24 volumio volumio[2756]: INFO:librespot_core::session: Country: "NL"
`



Trying a reboot to see if it sticks.

**Comment by Meistrovich** [comment]
For me its not working,
When I specify my spotify credentials in volumio I dont even see at the devices in spotify.
When it is set for multiple devices, I see the device in spotify but it doesnt connect to it.
Maybe I miss something.
I am quite noob yet, and dont know how to update apps from outside of volumio as balbuze asked for.

**Comment by ashthespy** [comment]
@Squixx Do you by any chance use a Facebook login? I don't see a reason offhand for this behaviour form your logs. Any info would be appreciated!


@Meistrovich  you need to [update the plugin via ssh](https://github.com/balbuze/volumio-plugins/tree/master/plugins/music_service/volspotconnect2#to-install) or wait until the update is available in the plugin manager.

**Comment by Squixx** [comment]
Yes, I am using Facebook login. I'd rather not use any login volumio side though.

I'm running on a pi 1 b, (old, 256mb)

**Comment by ashthespy** [comment]
I unfortunately don't have a FB login to reproduce with, if you are open to debugging a bit, hop onto gitter!

**Comment by ashthespy** [comment]
@Squixx could you checkout if the mdns services is being broad casted by the device?
` dns-sd -B _spotify-connect._tcp` should list the available zeroconf connect devices.

**Comment by nanab** [comment]
I have the same problem. I have two devices one whit standard spotify login and one whit facebook login. On the one whit spotify login everything works ok, but the one whit facebook login doesn't find volumio.
I run the latest version of Volumio on a Raspberry zero. Your plugin is installed from git rep.

If i run dns-sd -B _spotify-connect._tcp from my computer it dosent list anything.

Anything i can do to help you debug this?

**Comment by ashthespy** [comment]
It looks like for some reason the mDNS server is not being seen from the android client.
Can try installing a [Zeroconfig browser](https://play.google.com/store/apps/details?id=com.melloware.zeroconf) and check if you have the `_Volumio` and `_Spotify-Connect` services visible? 
I will in the meantime try and figure out why this is acting up for some people and not for all!

**Comment by nidomiro** [comment]
The Screenshot shows the state with "Multi-User Device" enabled. `_Spotify-Connect` is not visible.

![screenshot_20180415-110738](https://user-images.githubusercontent.com/15079172/38776858-1f91536a-409e-11e8-9252-c310f90399f7.png)

**Comment by ashthespy** [comment]
Thanks for the confirmation. Now to figure out why this is specific to Android. 
I just noticed that even though volspotconnect2 works fine for me on my Android, it takes a minute or two before this Zeroconfiguration app sees the `_Spotify-Connect` entry. Unfortunately, I don't have a ready solution - will investigate though!

**Comment by nidomiro** [comment]
Thanks :) For me it is ok to disable the "Multi-User Device" for now, because I'm the only user.

If you need any help, please let me know

**Comment by ashthespy** [comment]
Quick question - Do you by any chance have the snapcast plugin installed?

**Comment by nidomiro** [comment]
On my Volumio PI I have the following plugins installed:

- Spotify (I cannot Remove or update it. Every action is resulting in an unknown error)
- Volumio Spotify Connect2

Maby I'll do a clean install.

**Comment by ashthespy** [comment]
Just witnessed something:
Desktop client (WIn10 -1803) did not see the Multiuser device, but my Android client did. Enabling it on the Android, immateriality made it pop up on my desktop client as well. 
Need to dig a bit.

**Comment by nidomiro** [comment]
I had something similar the other way around.
I guess if you connect to a Device, it's registered to Spotify and other clients connected to the same account can pull that info.

**Comment by mdigit5** [comment]
I realize that this is a very old post... not trying to necropost or anything, but I experienced the same (or similar problem) and it was not related to the Volspotconnect2 plugin at all. Sharing this info to potentially help others.

**TL;DR**

1). It was all about my Wireless Access Point blocking multicast/broadcast traffic – even for hosts on the same network segment!  Read on if you think this could be your problem too.

2). I discovered a very nice app to help discover MDNS/Bonjour/Zeroconf traffic on the network called HE.NET – Network Tools (Hurricane Electric Tools). Installed on my phone (phone on the same wireless network as Volumio) I confirmed that there was absolutely no multicast traffic… which led me to look at the config on my router and Wireless AP and ultimately solve the problem.

--

First off, what is the Problem? With the Volspotconnect2 plugin correctly installed I could only make it work in single user mode - i.e. Multi-User Device: Off + a valid Spotify premium username and password specified directly in the plugin settings.

By "work" I mean with this configuration, if the same user (as defined in the plugin authentication settings) was logged into the Spotify app on their phone (or wherever), they could see and select Volumio as a Spotify Connect device and everything worked as expected. Great! But this only works for a single user, and **it gave a false sense of what was actually happening / why it was working.**

Trying to get Multi-User mode to work from here I did the following:

Set Multi-User Device to On and left the credentials in the config: This didn’t seem to change anything. i.e. The plugin still authenticated to Spotify as the single defined user and things continued to work as described above. But still, no other users on the network could see Volumio as a Spotify Connect device.

Set Multi-User Device to On AND cleared the credentials: This resulted in Volumio not being seen at all – not by any user. i.e. At this point, Multi-User mode is truly enabled, but it's completely "broken".

Why?

The problem was in my WiFi Access Point. It was configured to block all multicast and broadcast traffic (The reason is because this traffic can cause performance problems on the AP- so if it’s not needed, it’s disabled). But, this traffic is necessary for other WiFi clients on the same network to see Volumio as a Spotify Connect enabled device (Volumio essentially sends info out to all other devices on the network announcing to them that it exists – and this announcement was blocked). As soon as I enabled the multicast/broadcast/mdns traffic in the AP, everything started working perfectly.

You can stop here... but, if you want to understand why I could see Volumio as a Spotify Connect device from my phone when Volspotconnect2 was configured in Single User mode (even when the local multicast/broadcast traffic was blocked), keep reading….

When Volspotconnect2 was configured in single user mode with my Spotify premium username and password, it connected to my Spotify account and essentially registered Volumio as an available Spotify Connect device (i.e. it registered the Volumio device with the Spotify service in the cloud under my specific Spotify account). Then, from any other Spotify app (like on my phone) logged into Spotify with the same account, Spotify (not the multicast/broadcast info on the local network) told the app about Volumio.  This information provided directly from the Spotify service is what allowed the app on my phone to “see” Volumio as a valid Spotify Connect target.

This also explains why any other user attempting to “see” Volumio as a valid Spotify Connect target on the local network could not see it even when it was working for my account. i.e. Volumio was not authenticated / registered to their Spotify account, so Spotify did not know about it / could not tell the app about it.

Anyway… I know it’s a long-drawn-out explanation… but, this took me a few hours to figure out, so I figured I’d try to share the knowledge with anyone else running into this.
